### ISSUE TITLE
REGRESSION: 1.21 Problem with Menu Links

### ISSUE THREAD

**Issue by stpaultim** [backdrop/backdrop-issues#5460]
## Description of the bug

This problem has been reported in the forum: https://forum.backdropcms.org/forum/troubles-121

"pb with new version on my menu : each "update.php" multiplies menu links on my main menu; where I had 10 links described, I end up with 20 or more links that I have to manually deactivate and I have to do it every "update.php" because they reactivate themselves; I have no idea how to remove these links.. the site becomes unmanageable"

## Steps To Reproduce

I have yet to reproduce this problem, but wanted to make sure it gets into the core queue quickly in case anyone has any insights.

**Comment by herbdool** [comment]
I posted in the forum comments that I had attempted to recreate the issue on a fresh Backdrop site and was unable to. I created a fresh install of 1.20.3, created some dummy content with links in the main menu. Then upgraded the code to 1.21.0 and ran update.php. Everything appeared as normal.

**Comment by docwilmot** [comment]
This is happening to me. On a site, I now have 16 copies of `user/login` in my db in the `menu_links` table, and a new one is created with each cache clear.

**Comment by docwilmot** [comment]
Strangely the dupes in the DB are all marked as menu callbacks, so theyre not appearing in the site menus

**Comment by bugfolder** [comment]
I've seen this also on my test site, and the dupe is `user/login`. If I delete all of the recent `user/login` menu link entries (so that the last entry is in the table is something else), then re-clear the cache, the added entry is still `user/login`.

**Comment by docwilmot** [comment]
Interestingly I noticed this because my user menu suddenly had two copies each of login and logout. I disabled the extras.

**Comment by argiepiano** [comment]
In a clean install with 1.21.0, I can confirm that a `user/login` entry is added to `menu_links` every time the cache is cleared. The parent link ID `plid` for all but one of those entries is 2, which is the `user-menu` (Account menu). The first entry has a plid of 0. All the entries whos plid is 2 show `system` as in the module column. The one with plid 0 has `menu` in the module column.  

Additionally, only one is shown in the Account menu - the one with plid 0 (in my install thats mlid 329, the first entry in the list)

**Comment by argiepiano** [comment]
One more clue -  there is a link path `user/login` with mlid 13. The duplicates are all added at the end of the table, with mlid > 329 etc.

I've check an old test site that uses an earlier version of Backdrop. This duplication is not seen there.

**Comment by herbdool** [comment]
This might be the first place to look https://github.com/backdrop/backdrop-issues/issues/5389. Not sure

**Comment by docwilmot** [comment]
> This might be the first place to look #5389. Not sure

Quite likely. Maybe this isnt the same as the original post? I noticed I do have two copies of `user/logout` in the user menu as well., so still possible. @hosef?

**Comment by argiepiano** [comment]
> This might be the first place to look #5389. Not sure

Yes. I've started tracing the issue, and it seems like one potentially problematic line is https://github.com/backdrop/backdrop/blob/f8d209a31cfe74841924be8ebb8c098fe47fceed/core/includes/menu.inc#L3017

In a clean install, the first anomalous entry of user/login is mlid 329, which, interestingly, has a module column set to `menu` instead of `system`. When you rebuild the menu links with `function _menu_navigation_links_rebuild()`, the function loops through the menu_links cache - it finds that particular entry with mlid 329, and checks the `module` in that line above . Since module is set to `menu` instead of `system` it proceeds to create a new menu_links entry.

My research is still very preliminary. I don't know when/why the entry with mlid 329 is created (it's the first duplicate, and it has a plid of 0 instead of 2). But this can provide some clues.

**Comment by argiepiano** [comment]
My hypothesis in my comment above about the potentially problematic line was not completely accurate. 

However,  I have found when and why the anomalous second entry of `user/login` is created.  I believe this is at the root of the problem. This "first" duplicate entry (which generates the rest of the duplicates on cache clearing) is created by the invocation of `menu_link_save()` in line 433 of `standard.install`. 

I may be wrong, I have a very strong hunch that the problem is with some of the modifications in `menu_link_save()` introduced in #5389.  I don't have time to check on this further tonight, but I believe we should be looking at this particular invocation of `menu_link_save()` with a very skimpy array created in line 433. I'll be able to check on this further tomorrow.

**Comment by findlabnet** [comment]
Confirmed bug on all 1.21.0 site. 
`user/login` entry is added to `menu_links` every time the cache is cleared. 
The `plid` for all but first of those entries is 2, the first entry has a `plid` of 0. 
All new marked as menu callbacks, and not appearing in the site real menus.

**Comment by argiepiano** [comment]
OK, one last post and I'm done tonight. 

I need to scratch my hunches and hypotheses above. I've been tracing the installation with Backdrop 1.20.4. In fact there ARE two entries of `user/login` created there in the table menu_links during installation of 1.20.4. So that call to menu_link_save in standard.install is not problematic.

The problem may still be located in  _menu_navigation_links_rebuild, but it occurs after installation, when the menus are rebuilt.

**Comment by klonos** [comment]
This is now a confirmed and serious issue. Setting the milestone to 1.21.1.

**Comment by findlabnet** [comment]
Backdrop 1.19.3 (outdated dev. site) - two entries of `user/login` - one normal with `mlid` 315, second hidden (callback) with `mlid` 13.

**Comment by indigoxela** [comment]
Confirmed. The duplicated menu links aren't visible, but in table menu_links for every cache clear another item appears.

I did a step-by-step upgrade from 1.20.3 -> 1.20.4 -> 1.21 pre. The problem starts with 1.21.0-preview.

**Comment by indigoxela** [comment]
Yes, it's the menu_rebuild optimization causing the problem. Reverting the change on a dev web stopped the duplication.

I created a [PR](https://github.com/backdrop/backdrop/pull/3912) that simply reverts the commit, in case we do not find a better solution.
It also seems important that we don't miss this issue, when the next bugfix release goes out.

**Comment by indigoxela** [comment]
Can anyone confirm that this bug really only affects the link_path `user/login` - that's the only faulty duplicates I found. No other path seems affected.

**Comment by indigoxela** [comment]
Usually I don't use uppercase, but I wanted to make this issue more visible in the milestone list. :wink:

**Comment by herbdool** [comment]
I think we might apply the patch on our client sites. One client just reported not being able to next menu items. I'm not sure if it's related. If we don't figure out a fix in the next day or so then I'll test/review @indigoxela's PR.

**Comment by olafgrabienski** [comment]
> not being able to next menu items

Not being able to *what* ?-)

**Comment by indigoxela** [comment]
> One client just reported not being able to nest menu items

I can't reproduce that.

And I also can't reproduce the duplication of **visible** menu items as reported in the forum. That makes me suspect, that the update hook might not catch all duplicated menu items.

**Comment by herbdool** [comment]
I haven't been able to recreate the nesting issue locally but it was a fresh 1.21.0 site not upgraded.

**Comment by argiepiano** [comment]
@indigoxela thanks for the patch. I can confirm that the patch stops the duplications of user/login in the menu_links table.

However, it doesn't remove the duplicate user/login from the table.

**Comment by indigoxela** [comment]
> However, it doesn't remove the duplicate user/login from the table.

Did you run update.php? Anything in dblog? If that's all good, can you provide a list of the remaining user/login records in table menu_links? It should be two or some more.

**Comment by argiepiano** [comment]
> And I also can't reproduce the duplication of **visible** menu items as reported in the forum. That makes me suspect, that the update hook might not catch all duplicated menu items.

This IS a very difficult issue to reproduce. I've tried a couple of things: adding links of existing menu_router items manually to menu, and also running this in Devel:

```php
  $item = array(
    'link_title' => st('Logout'),
    'link_path' => 'user/logout',
    'menu_name' => 'main-menu',
'weight' => -1
  );
  menu_link_save($item);
```

The above adds Logout item to Navigation menu normally, but the cache clears do not duplicate it in the menu_links  table like it did with the user/login that was added during installation.

Also, those reports, especially phillipegs in the forum. I'm wondering if the issue is an interaction of the problem we are seeing with 1.21.0 AND a contrib module that somehow is adding links to menu_links programmatically. 

For me the first step is to troubleshoot the issue with user/login. That may fix the rest. 

Also, it's great to have the "revert" patch that @indigoxela created, but that will be hard to apply for non-developers. So, we may consider an emergency release of Backdrop that includes @indigoxela patch for now?

**Comment by indigoxela** [comment]
It's a true puzzle, for sure.

> ... do not duplicate it in the menu_links table like it did with the user/login that was added during installation

I'm not sure if that's correct. The menu item that gets added in the standard profile installation is a *regular menu item* - it's item has "menu" in the "module" column.
The one that gets duplicated is the menu callback record (module = system, hidden = -1).

**Comment by argiepiano** [comment]
> The one that gets duplicated is the menu callback record (module = system, hidden = -1

You are right.

**Comment by quicksketch** [comment]
I'm in favor of reverting, as @indigoxela has done in https://github.com/backdrop/backdrop/pull/3912.

Regarding the update hook to clean up the mess, is the problem _only_ with `user/login`? Or are there other paths for which this occurs too?

**Comment by hosef** [comment]
@quicksketch I am currently trying to debug the behavior. On my local environment, there are 2 entries that seem to get duplicated every time caches are cleared. `user/login` from `user-menu` and `admin/config` from `management`.

**Comment by herbdool** [comment]
My coworker applied the patch to revert this on a client's site and it seems to have fixed the issue. For the record this was a Drupal 7 site that we upgraded to Backdrop a couple years ago. And this is the report from the client of their issue:

> We have noticed today that after editing and saving a page, the page's menu item is no longer nested properly. We attempted to update the order of menu items via page https://REDACTED/admin/structure/menu but the change didn't stick.

**Comment by argiepiano** [comment]
@hosef I spent a bit of time just now and last night. I think I have found the source of the problem.

The new menu_links  cache  used by `_menu_navigation_links_rebuild` is keyed by path. The problem is that  sometimes there may be more than one entry of the same path in `menu_links`, as we have seen in comments above (this is normal), for example when you insert a link into a menu of an existing menu_router item.

In the case of `user/login`, there are two "normal" entries in menu_links, one created by hook_menu in `user.module` and one link inserted during installation in `standard.install`. The two entries are different in one important aspect: one of them is marked as `system` and the other as `menu` in the module column.

The issue here is that the cache only stores the one with the `menu` column, and therefore the if statement in line 3016 fails, since it's looking for `system` links. Therefore, later in this function, a new entry for the `user/login` entry with `system` is created.

So, there is no easy solution. Perhaps the cache can be modified so that it's keyed with a string created with the path AND the module

Again, all of the above are preliminary observations and may not be completely accurate.

I hope this helps.

**Comment by argiepiano** [comment]
> Regarding the update hook to clean up the mess, is the problem _only_ with `user/login`? Or are there other paths for which this occurs too?

According to a report in the forum, the duplications are happening to multiple items. However, people in this thread have not been able to reproduce that. The patch provided by @indigoxela stops the duplication  - duplications beyond user/login can probably be removed manually from the menus?

**Comment by hosef** [comment]
> The new menu_links cache used by `_menu_navigation_links_rebuild` is keyed by path. The problem is that sometimes there may be more than one entry of the same path in `menu_links`, as we have seen in comments above (this is normal), for example when you insert a link into a menu of an existing menu_router item.

Interesting. I just checked on my local and I had 2 links for `admin/config` as well, so that explains why that is being duplicated on my local site.

**Comment by argiepiano** [comment]
When building the menu_link_cache for the first time, THIS is problematic (line 3117 in `_menu_links_cache()`:

```php
  if (empty($menu_links) || $reset) {
    $menu_links = db_select('menu_links')
      ->fields('menu_links')
      ->orderBy('depth', 'DESC')
      ->execute()->fetchAllAssoc('link_path');
  }
```

This clearly removes items with the same path that have  multiple entries in the table, as this keys all entries by `link_path`  with `fetchAllAssoc`

**Comment by argiepiano** [comment]
So, one possible solution is as I mentioned above, to preserve those "normal" duplications in the cache by keying it in a different way, for example `link_path:module` or something like that, and modifying the checks in `_menu_navigation_links_rebuild()` to check for that key instead of just `link_path`.

**Comment by argiepiano** [comment]
(As a related side note [or side question]: why are we not using the Cache API provided by Backdrop for this cache?)

**Comment by hosef** [comment]
So, I think that it might be better to revert my original change and then rework it to use `mlid` instead of `link_path` for the array index later. That should fix the issue because `mlid` is a unique key, but it may take some time to get everything worked out and also come up with tests to make sure this doesn't happen again.

**Comment by herbdool** [comment]
RTBC

**Comment by argiepiano** [comment]
One more piece of information about this bug that may help explain the issue with nested menus. 

This information may all be a "moot point" as it was decided to revert back this feature. But perhaps it may be helpful in the future.

- When building the menu_links_cache, `_menu_links_cache()` queries the menu_links table and sorts all the entries by descending `depth`. Depth is an indication of how deep in the tree this link is. For example a depth of 2 means that there is another item "above" in the menu. So items with "2" will appear higher that "1" in the results
- As we saw above, this query fetches the items in menu_links keyed by `link_path`. This effectively means that, when there are multiple entries of a single link_path (e.g. user/login), only one menu_link item - the last one in the results
- In the case of `user/login`, the entry created with hook_menu in user.module (which has a `system` in its module column) has a depth of 2, while the one created by standard.install (which has `menu` in the module column) has a depth of 1. 
- This means that the query will only pick the lower one (the last one it finds in the sorted results), in this case, ALWAYS the one created by standard.install with depth of "1" and `menu` as its module column
- And this creates a problem, as the function `_menu_navigation_links_rebuild` will trigger a new save for the link since it thinks that a link for the "hook_menu" user/login doesn't exists in the menu_links table.  Thus the infinite duplications

This also explains why we are not seeing a similar issue with other hook_menu items like `user/logout` after they are  manually added to any of the menus. In fact if you add user/logout to the top of the Navigation menu, you will see ONE duplication (i.e. 3 entries in the table), and then the duplication stops. This is because both the "hook_menu" user/logout and the one added manually have both the same depth (1). Which means that the query will at first pick the one added manually (the last one added). After that, It will create ONE duplication of the one created by hook_menu, which, will effectively now become the last one in the cache query. From now on the query will always pick the "hook_menu" user/logout entry for the cache, effectively avoiding the continuous duplication.

This may also explain some situations when the nesting is lost, since, for items that exists in multiple menus, the cache will only pick the one with the lowest depth

**Comment by herbdool** [comment]
Good detective work! I'm going to test items in multiple nestings. That might be it.

**Comment by quicksketch** [comment]
We discussed this during the weekly call today and we're going to do the following:

- The current PR prevents further problems and will help all sites that have not yet upgraded. It also runs an update hook to revert the most common database problem of user/login being duplicated, but is not comprehensive.
- If a full fix to all menu links in the database is available by tomorrow we'll use that PR. Otherwise we'll continue with a more comprehensive fix in a separate issue.

Either way, we'll use the available solution to fix this problem as soon as possible. We'll be making a 1.21.1 release tomorrow, Friday January 21.

**Comment by indigoxela** [comment]
> It also runs an update hook to revert the most common database problem of user/login being duplicated, but is not comprehensive.

@quicksketch What would you expect to make it more comprehensive?

The update hook is as cautious as possible and only deletes entries, of which we know how duplication happened, and that all sites are affected. Deleting more, especially visible (regular) menu items (with module set to "menu") seems way too dangerous to me. More than one regular menu items with the same path are valid, for instance for different languages or in different menus. I wouldn't dare to touch them, but those can get deleted manually, as they are accessible via UI.

And special thanks to @argiepiano and @hosef - great sleuthing!

**Comment by indigoxela** [comment]
Aaaand here's a way to reproduce the duplication of *visible* regular menu items:

- create a views page (content doesn't matter) and add a menu link via views ui
- add another menu item to the same view via menu ui
- flush menu cache

Just for later testing of the overhauled optimization. I think, here we have consensus to revert that commit.

**Comment by indigoxela** [comment]
> ... but those can get deleted manually, as they are accessible via UI.

Unfortunately that's not entirely correct. If following my steps to reproduce, the link from views gets created with "module = system", and that one gets duplicated. That link can not get deleted via menu UI, and even if one removes the link from view via views UI, the duplicates persist, only the valid one gets removed.

**Comment by indigoxela** [comment]
As we're now able to reproduce additional duplications (more paths than just "user/login" are affected), I created an [alternate PR](https://github.com/backdrop/backdrop/pull/3914), which does more cleanup.

Please take a *close* look at that extended menu_update_1005() - is it really safe to do that? Besides the update hook, the second PR is identical to the first.

**Comment by herbdool** [comment]
@indigoxela I'll test - thank you for your work on this.

I made a mistake of putting the initial patch on backdrop-pantheon. I didn't realize it had an update hook in there. Could make the hook `menu_update_1006` to cover up my mistake (since they can't be deleted via the UI)?

**Comment by indigoxela** [comment]
> I made a mistake of putting the initial patch on backdrop-pantheon...

Ooops, I don't know how this normally gets solved. If you make it 1006 the next one would fail.

Many thanks for testing.

**Comment by herbdool** [comment]
I tested and reviewed https://github.com/backdrop/backdrop/pull/3914. It looks good.

Regarding, changing the numbering to `1006`, I'm not sure why the next one would fail. Wouldn't we just make the next one `menu_update_1007`? We would just skip 1005, or make it an empty function with a note. If that's not a good idea I'll have to provide some tips for Pantheon users on how to get rid of duplicates via some database calls. Not clear how many Pantheon sites there are.

**Comment by indigoxela** [comment]
Ah, now I get it, you want us to make it 1006 *here in this PR* and add a "Update removed" hint to 1005. Sure, technical possible, but I'm not the one to decide.

**Comment by argiepiano** [comment]
I have reproduced the views-generated menu links duplication problem following @indigoxela's directions above.

I've tried applying the second patch (https://github.com/backdrop/backdrop/pull/3914) to my test site and have found the following issue (which I believe was addressed above):

- Yesterday I had applied the previous patch created by @indigoxela  - 3912 (https://github.com/backdrop/backdrop/pull/3912) which contained `menu_update_1005()`. I ran update.php with that patch installed.
- Then today, for more testing, I reinstalled the unpatched 1.21.0 WITHOUT modifying my database at all (basically moved the core files of the unpatched 1.21.0 download into my site's base directory).
- Then I applied the new patch, 3914,  https://github.com/backdrop/backdrop/pull/3914, but when I run update.php, the new update hook does not run, since it's also named `update_1005` in the new patch, and Backdrop thinks it's already applied (from yesterday's update.php)

This is a possible scenario for those who rushed to apply @indigoxela's first patch from yesterday. So, yes, we'll need to change the`hook_update_` of the second PR to 1006, and include a dummy one 1005 in it, I think.

I just did that manually, and the patch worked as expected. Looks good to me!

@indigoxela thanks so much for doing all of this so quickly!

**Comment by argiepiano** [comment]
I'm marking this as "needs work" so that the update hook is fixed.

**Comment by quicksketch** [comment]
I'm in favor of making an empty `menu_update_1005()` function and renaming to `menu_update_1006()` in the second PR. Pantheon is a semi-official upstream since it's part of the actual release process.

Still not clear if we'll use the second PR or the first one, so keeping both as options is a good idea.

**Comment by argiepiano** [comment]
The second PR seems better than the first in my opinion, as it clears out all duplications, not just user/login. I feel it's safe, since it only focuses on those entries marked as `system` in the module column. It's my understanding that normally those entries appear only once in the menu_links table. These entries are those inserted there for menu_router items created with hook_menu  (this includes the ones created by choosing "Menu" in Page Settings in the Views UI). 

All manual additions of links to menus, as well as additions made in code by invoking menu_link_save() like the ones in standard.install have a value of `menu` in the module column. Therefore the update hook provided by @indigoxela  will not touch them AFAIK.

**Comment by argiepiano** [comment]
Of course, this is such a delicate issue that we really would like input from others. Maybe @hosef can also chime in?

**Comment by hosef** [comment]
I tested the second PR and it seems to work as expected. I think it's a good idea to clean up more links than just the `user/login` ones.

**Comment by herbdool** [comment]
I updated the hook numbers and added an empty update hook - hope that's okay @indigoxela. But there's still a question from @quicksketch in the PR so can't mark the second PR as RTBC yet.

**Comment by quicksketch** [comment]
I checked on my own question: Do other modules that use `hook_menu()` result in `system` being the module name in the `menu_links` table? The answer is yes. So @indigoxela's fix looks correct to me and should be comprehensive of duplicate-created links.

I tested this locally on an intentionally broken 1.21.0 site by creating a view with a normal menu item, then manually creating a menu link at the same path, and clearing the cache. I ended up getting new menu entries after the cache clear (bringing total to 3 menu links at the same path). I also created intentional duplicate menu links that are _not_ created by a module to confirm they don't get accidentally deleted.

Before running update:
```
MariaDB [db]> select menu_name, mlid, plid, link_path, link_title, langcode, module, COUNT(mlid) as count from menu_links GROUP BY link_path HAVING count > 1;
+-----------+------+------+--------------+------------+----------+--------+-------+
| menu_name | mlid | plid | link_path    | link_title | langcode | module | count |
+-----------+------+------+--------------+------------+----------+--------+-------+
| main-menu |  351 |    0 | listing-test | Listing    | und      | system |     3 |
| main-menu |  335 |    0 | node/1       | Test       | und      | menu   |     9 |
| user-menu |   13 |    2 | user/login   | Log in     | und      | system |    25 |
+-----------+------+------+--------------+------------+----------+--------+-------+
```

After running update hook, duplicates are removed:
```
MariaDB [db]> select menu_name, mlid, plid, link_path, link_title, langcode, module, COUNT(mlid) as count from menu_links GROUP BY link_path HAVING count > 1;
+-----------+------+------+--------------+------------+----------+--------+-------+
| menu_name | mlid | plid | link_path    | link_title | langcode | module | count |
+-----------+------+------+--------------+------------+----------+--------+-------+
| main-menu |  351 |    0 | listing-test | Listing    | und      | system |     2 |
| main-menu |  335 |    0 | node/1       | Test       | und      | menu   |     9 |
| user-menu |   13 |    2 | user/login   | Log in     | und      | system |     2 |
+-----------+------+------+--------------+------------+----------+--------+-------+
```

I tested this with the empty update hook in the latest PR. This second PR looks safe and ready to me.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/3914 into 1.x and 1.21.x. We'll be making the 1.21.1 release shortly today.
